tls/s2n_tls13_key_schedule.c

```diff
@@ -0,0 +1,327 @@
+        K_send(conn, S2N_APPLICATION_SECRET);
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#appendix-A.2
+     *# here                  +--------+--------+
+     *#              No 0-RTT |                 | 0-RTT
+     *#                       |                 |
+     *#   K_recv = handshake  |                 | K_recv = early data
+     */
+        if (WITH_EARLY_DATA(conn)) {
+            K_recv(conn, S2N_EARLY_SECRET);
+        } else {
+            K_recv(conn, S2N_HANDSHAKE_SECRET);
+        }
+    }
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#appendix-A.2
+     *# [Skip decrypt errors] |    +------> WAIT_EOED -+
+     *#                       |    |       Recv |      | Recv EndOfEarlyData
+     *#                       |    | early data |      | K_recv = handshake
+     *#                       |    +------------+      |
+     */
+    if (message_type == END_OF_EARLY_DATA) {
+        K_recv(conn, S2N_HANDSHAKE_SECRET);
+    }
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#appendix-A.2
+     *#                       |                        |
+     *#                       +> WAIT_FLIGHT2 <--------+
+     *#                                |
+     *#                       +--------+--------+
+     *#               No auth |                 | Client auth
+     *#                       |                 |
+     *#                       |                 v
+     *#                       |             WAIT_CERT
+     *#                       |        Recv |       | Recv Certificate
+     *#                       |       empty |       v
+     *#                       | Certificate |    WAIT_CV
+     *#                       |             |       | Recv
+     *#                       |             v       | CertificateVerify
+     *#                       +-> WAIT_FINISHED <---+
+     *#                                | Recv Finished
+     *#                                | K_recv = application
+     */
+    if (message_type == CLIENT_FINISHED) {
+        K_recv(conn, S2N_APPLICATION_SECRET);
+    }
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#appendix-A.2
+     *#                                v
+     *#                            CONNECTED
+     */
+    return S2N_RESULT_OK;
+}
+
+s2n_result (*key_schedules[])(struct s2n_connection*) = {
+    [S2N_CLIENT] = &s2n_client_key_schedule,
+    [S2N_SERVER] = &s2n_server_key_schedule,
+};
+
+S2N_RESULT s2n_tls13_key_schedule_update(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+    if (s2n_connection_get_protocol_version(conn) < S2N_TLS13) {
+        return S2N_RESULT_OK;
+    }
+    RESULT_ENSURE_REF(key_schedules[conn->mode]);
+    RESULT_GUARD(key_schedules[conn->mode](conn));
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_tls13_key_schedule_reset(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+    conn->client = &conn->initial;
+    conn->server = &conn->initial;
+    conn->secrets.tls13.secrets_count = 0;
```

Comment:
I am not understanding of what the purpose of secrets_count is. Can you explain?